src/hardware/dac.rs

```diff
@@ -0,0 +1,279 @@
+///! Thermostat DAC/TEC driver
+///!
+///! This file contains all of the drivers to convert an 18 bit word to an analog current.
+///! On Thermostat this used the ad5680 DAC and the MAX1968 TEC driver. The (analog voltage)
+///! max output voltages/current settings are driven by PWMs of the STM32.
+use num_enum::TryFromPrimitive;
+
+use super::hal::{
+    gpio::{
+        gpiob::*, gpioc::*, gpiod::*, gpioe::*, gpiog::*, Alternate, Output, PushPull, AF1, AF2,
+        AF6,
+    },
+    hal::{blocking::spi::Transfer, digital::v2::OutputPin, PwmPin},
+    prelude::*,
+    pwm::{ActiveHigh, ComplementaryDisabled, ComplementaryImpossible, C1, C2, C3, C4},
+    rcc::{rec, CoreClocks},
+    spi::{Enabled, NoMiso, Spi, MODE_1},
+    stm32::{SPI3, TIM1, TIM3, TIM4},
+    time::{MegaHertz, U32Ext},
+};
+
+use super::unit_conversion::{i_to_dac, i_to_pwm, v_to_pwm};
+
+pub const SPI_CLOCK: MegaHertz = MegaHertz(30); // DAC SPI clock speed
+pub const MAX_VALUE: u32 = 0x3FFFF; // Maximum DAC output value
+pub const F_PWM: u32 = 20; // PWM freq in kHz
```

Comment:
Why that value? Should be in argued in relation to the filter order, 3dB bandwidth and the desired accuracy (ripple due to the filter vs resolution on the PWM timer).

Reply:
That was what the m-labs thermostat-2ch code used. I'll check with the filter.

Reply:
Cutoff is 160Hz, 20kHz is ~80dB down. If we give 10% slack for the limits anyways I don't see the need to optimize this with the timer resolution. Should I?

Reply:
No. Just document the choice with those numbers.